sessiontxn/legacy/provider.go

```diff
@@ -122,6 +128,82 @@ func (p *SimpleTxnContextProvider) OnStmtStart(ctx context.Context) error {
 	return nil
 }
 
+// OnStmtErrorForNextAction is the hook that should be called when a new statement get an error
+func (p *SimpleTxnContextProvider) OnStmtErrorForNextAction(point sessiontxn.StmtErrorHandlePoint, err error) (sessiontxn.StmtErrorAction, error) {
+	switch point {
+	case sessiontxn.StmtErrAfterQuery:
+		return p.handleAfterQueryError(err)
+	case sessiontxn.StmtErrAfterLock:
+		return p.handleAfterLockError(err)
+	default:
+		return sessiontxn.NoIdea()
+	}
+}
+
+func (p *SimpleTxnContextProvider) handleAfterQueryError(queryErr error) (sessiontxn.StmtErrorAction, error) {
+	if p.Sctx.GetSessionVars().IsRcCheckTsRetryable(queryErr) {
+		logutil.Logger(p.Ctx).Info("RC read with ts checking has failed, retry RC read",
+			zap.String("sql", p.Sctx.GetSessionVars().StmtCtx.OriginalSQL))
+		return sessiontxn.RetryReady()
+	}
+	return sessiontxn.NoIdea()
+}
+
+func (p *SimpleTxnContextProvider) handleAfterLockError(lockErr error) (sessiontxn.StmtErrorAction, error) {
```

Comment:
This is now for `pessimistic` lock error only, maybe change the name to `AfterPessimisticLockError` or `PessimisticLockError`？

Reply:
done